Cycle in media type definitions is not properly handled

## Expected Behavior
Generated code with a collection of child elements
## Actual Behavior
Stack overflow
## Steps to Reproduce the Problem (including complete and simple design if relevant)

My scheme:
```go
// Recipe is the recipe resource media type.
var Recipe = MediaType("application/recipe+json", func() {
	Description("Recipe")
	ContentType("application/json")
	Attributes(func() {
		Attribute("id", UUID, "UUID of Recipe", func() {
			Example("cb241e3e-4e60-46fc-8ef9-2e46261cc378")
		})
		Attribute("parentId", UUID, "UUID of parent recipe", func() {
			Example("152064a6-9a5e-4791-803b-05a72b8cbf50")
		})
		Attribute("modifications", CollectionOf("application/recipe"))
		Attribute("name")

		Attribute("createdAt", DateTime, "Date of creation")
		Attribute("updatedAt", DateTime, "Date of last update")
		Required("id", "parentId", "name")
		Required("createdAt")
	})

	View("default", func() {
		Attribute("id")
		Attribute("parentId")
		Attribute("name")
		Attribute("modifications")
		Attribute("createdAt")
		Attribute("updatedAt")
	})

	View("tiny", func() {
		Attribute("id")
		Attribute("parentId")
		Attribute("createdAt")
	})
})
```
stack overflow:
```
exit status 2
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x145b9a3, 0xe)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/panic.go:616 +0x81
runtime.newstack()
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/stack.go:1054 +0x71f
runtime.morestack()
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/asm_amd64.s:480 +0x89

goroutine 1 [running]:
runtime.mapiternext(0xc440282400)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/hashmap.go:740 +0x563 fp=0xc440282380 sp=0xc440282378 pc=0x100a783
runtime.mapiterinit(0x142bc00, 0xc42022b650, 0xc440282400)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/hashmap.go:737 +0x1f1 fp=0xc4402823a8 sp=0xc440282380 pc=0x100a131
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc4200ada80, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:635 +0x2db fp=0xc440282470 sp=0xc4402823a8 pc=0x12f917b
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc4200adec0, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:616 +0xe4 fp=0xc440282538 sp=0xc440282470 pc=0x12f8f84
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc4200adb40, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:636 +0x33a fp=0xc440282600 sp=0xc440282538 pc=0x12f91da
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc42028a000, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:616 +0xe4 fp=0xc4402826c8 sp=0xc440282600 pc=0x12f8f84
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc4200adb40, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:636 +0x33a fp=0xc440282790 sp=0xc4402826c8 pc=0x12f91da
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc42028a000, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:616 +0xe4 fp=0xc440282858 sp=0xc440282790 pc=0x12f8f84
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc4200adb40, 0x0, 0x0, 0x0, 0x0)
	/Users/admin/go/src/github.com/goadesign/goa/design/types.go:636 +0x33a fp=0xc440282920 sp=0xc440282858 pc=0x12f91da
github.com/goadesign/goa/design.hasFile(0x14bbfc0, 0xc42028a000, 0x0, 0x0, 0x0, 0x0)
...additional frames elided...

goroutine 21 [syscall]:
os/signal.signal_recv(0xc420042fc8)
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/sigqueue.go:139 +0xa7
os/signal.loop()
	/usr/local/Cellar/go/1.10.3/libexec/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/usr/local/Cellar/go/1.10.3/libexec/src/os/signal/signal_unix.go:28 +0x41

make: *** [generate] Error 1
```

With `CollectionOf(Recipe)` I get 
```
../design/media_types.go:127:5: initialization loop
```
immediately.

## Specifications

  - goa Version: reported version is 1.3.1 but I think it's latest 1.4.0
  - Platform (Linux/Windows etc.): macOS

--- a/design/types.go
+++ b/design/types.go
@@ -605,7 +605,7 @@ func HasFile(dt DataType) bool {
 	return hasFile(dt, nil)
 }
 
-func hasFile(dt DataType, traversed []DataType) bool {
+func hasFile(dt DataType, seen map[string]struct{}) bool {
 	if dt == nil {
 		return false
 	}
@@ -613,32 +613,31 @@ func hasFile(dt DataType, traversed []DataType) bool {
 	case dt.IsPrimitive():
 		return dt.Kind() == FileKind
 	case dt.IsArray():
-		if hasFile(dt.ToArray().ElemType.Type, traversed) {
-			return true
+		et := dt.ToArray().ElemType.Type
+		if _, ok := seen[et.Name()]; ok {
+			return false
 		}
+		return hasFile(et, seen)
 	case dt.IsHash():
-		if hasFile(dt.ToHash().KeyType.Type, traversed) {
+		if hasFile(dt.ToHash().KeyType.Type, seen) {
 			return true
 		}
-		if hasFile(dt.ToHash().ElemType.Type, traversed) {
-			return true
-		}
-	case dt.IsObject() && dt.Kind() == UserTypeKind:
-		for _, v := range traversed {
-			if dt == v {
-				return false
-			}
-		}
-		traversed = append(traversed, dt)
-		fallthrough
+		return hasFile(dt.ToHash().ElemType.Type, seen)
 	case dt.IsObject():
+		if _, ok := seen[dt.Name()]; ok {
+			return false
+		}
+		if seen == nil {
+			seen = make(map[string]struct{})
+		}
+		seen[dt.Name()] = struct{}{}
 		for _, att := range dt.ToObject() {
-			if hasFile(att.Type, traversed) {
+			if hasFile(att.Type, seen) {
 				return true
 			}
 		}
 	default:
-		panic("unknown type")
+		panic("unknown type") // bug
 	}
 	return false
 }

